coveo-functools/coveo_functools/flex/deserializer.py

```diff
@@ -55,32 +61,91 @@ def convert_kwargs_for_unpacking(dirty_kwargs: Dict[str, Any], *, hint: MetaHint
+) -> T:
     """This overload tricks mypy in passing typing annotations as types, such as List[str]."""
 
 
-def deserialize(value: Any, *, hint: Union[T, Type[T]]) -> T:
+def deserialize(
+    value: Any,
+    *,
+    hint: Union[T, Type[T]],
+    errors: ErrorBehavior = "deprecated",
+) -> T:
     """
     Deserializes a value based on the provided type hint:
 
     - If value is None, we return None. We don't validate the hint.
     - If hint is a builtin type, we blindly return the value we were given. We don't validate the value.
     - If hint is a custom class, the value must be a dictionary to "flex" into the class.
     - Hint may be a `Union[T, List[T]]`, in which case the value must be a dict or a list of them.
+
+    The `errors` argument controls the behavior when a value cannot be deserialized into the hint or annotation:
+        - 'ignore': the value is used as-is, without deserialization. errors will be logged.
+        - 'silent': behave like 'ignore' but don't log errors  # yolo
+        - 'raise': raise a PayloadMismatch exception.
+        - 'deprecated': (default) different situations yield different behaviors:
+            - type mismatch between value and hint (e.g.: cannot use a list to create a dict) would behave like 'ignore'
+            - value mismatch (e.g.: dict missing some arguments vs a class's __init__) would raise TypeError
+
+    Note that the `deprecated` option is currently the default, for legacy reasons. It will be removed because
+    it created an inconsistent behavior (see the examples below).
+
+    example #1: `range` requires the `stop` argument, which is missing:
+
+        >>> deserialize({}, hint=range, errors='raise')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0
+        >>> deserialize({}, hint=range, errors='deprecated')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0
+        >>> deserialize({}, hint=range, errors='ignore')
+        {}
+
+    example #2: the hint is a dict, but a list was provided:
+
+        >>> deserialize([], hint=dict, errors='raise')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        coveo_functools.exceptions.PayloadMismatch: I don't know how to fit <class 'list'> into <class 'dict'> of typing.Any
+        >>> deserialize([], hint=dict, errors='deprecated')
+        []
+        >>> deserialize([], hint=dict, errors='ignore')
+        []
     """
+    valid_error_modes = "raise", "ignore", "silent", "deprecated"
+    if errors not in valid_error_modes:
+        raise ValueError(f"'{errors=}' is not valid, {valid_error_modes=}")
+
+    if errors == "deprecated":
+        warnings.warn(
+            "Please specify the error behavior when calling `flex.deserialize`. "
+            "Recommended fix: specify `errors='raise'` to catch deserialization errors. "

```

Comment:
Another choice would have been to use something like an `enum` to represent the error mode. The trade-off I see are

**Good points of using `enum`**:
1. No typos are possibles (more precisely, typos would be caught at static analysis time by e.g. `mypy`).

**Good points of using a _string_**:
1. People need to import the `enum` to use it. Hence, the `Recommend fix` won't fit on a single line.

I think in the current context, I'd value more the `enum` than the string. However, you already implemented things with a string, and I'm not sure the benefit of using an `enum` is worth the time required to do the change. :thinking:

Reply:
I'm a BIG enum fan, but I don't want people to be forced to import 2 symbols from flex. 

For your information, I used the `Literal` type, so it's actually type safe! :)

Reply:
I missed that! :heart: